web/src/main/java/org/springframework/security/web/authentication/session/ConcurrentSessionControlAuthenticationStrategy.java

```diff
@@ -156,18 +158,12 @@ protected void allowableSessionsExceeded(List<SessionInformation> sessions,
 					"Maximum sessions of {0} for this principal exceeded"));
 		}
 
-		// Determine least recently used session, and mark it for invalidation
-		SessionInformation leastRecentlyUsed = null;
-
-		for (SessionInformation session : sessions) {
-			if ((leastRecentlyUsed == null)
-					|| session.getLastRequest()
-							.before(leastRecentlyUsed.getLastRequest())) {
-				leastRecentlyUsed = session;
-			}
+		// Determine least recently used sessions, and mark them for invalidation
+		sessions.sort(Comparator.comparing(SessionInformation::getLastRequest));
+		int maximumSessionsExceededBy = sessions.size() - allowableSessions + 1;
+		for (int i = 0; i < maximumSessionsExceededBy; i++) {
```

Comment:
`sessions` is defined as a `List` which means we need to treat this as though it could be any implementation of a `List`. If it were a `LinkedList`, then iterating over the `List` using `List.get(int)` is very inefficient.

Instead of using `List.get(int)` we should use the iterator (i.e. `for (SessionInformation session : sessions)`). A few options that use an Iterator are:

- keep a separate variable for determining which sessions should be expired and break out of the `for` loop if that is met. 
- produce [subList](https://docs.oracle.com/javase/8/docs/api/java/util/List.html#subList-int-int-) of sessions to expire and iterate over those

Reply:
@rwinch Thanks, I've updated it to use subList.